packages/core/src/transports/offline.ts

```diff
@@ -0,0 +1,146 @@
+import type { Envelope, InternalBaseTransportOptions, Transport, TransportMakeRequestResponse } from '@sentry/types';
+import { logger } from '@sentry/utils';
+
+export const START_DELAY = 5_000;
+const MAX_DELAY = 2_000_000_000;
+const DEFAULT_QUEUE_SIZE = 30;
+
+function wasRateLimited(result: TransportMakeRequestResponse): boolean {
+  return !!(result.headers && result.headers['x-sentry-rate-limits']);
+}
+
+type BeforeSendResponse = 'send' | 'queue' | 'drop';
+
+interface OfflineTransportOptions extends InternalBaseTransportOptions {
+  /**
+   * The maximum number of events to keep in the offline queue.
+   *
+   * Defaults: 30
+   */
+  maxQueueSize?: number;
+
+  /**
+   * Flush the offline queue shortly after startup.
+   *
+   * Defaults: false
+   */
+  flushAtStartup?: boolean;
+
+  /**
+   * Called when an event is queued .
+   */
+  eventQueued?: () => void;
+
+  /**
+   * Called before attempting to send an event to Sentry.
+   *
+   * Return 'send' to attempt to send the event.
+   * Return 'queue' to queue the event for sending later.
+   * Return 'drop' to drop the event.
+   */
+  beforeSend?: (request: Envelope) => BeforeSendResponse | Promise<BeforeSendResponse>;
```

Comment:
Yep, agreed, `beforeSend` is awful naming!

I was thinking that this hook could be used to control queuing behaviour but it can't actually do that fully since everything is automatically queued after failure. Let me rethink it...